bip-vaults.mediawiki

```diff
@@ -0,0 +1,757 @@
+<target-outputs-hash>
+<trigger-sPK>
+<trigger witness stack item 1> [...] <item n>
+</source>
+
+If the witness stack consists of fewer than four items the script MUST fail and
+terminate immediately.
+
+(Note: in practice, the witness stack will have included the other items necessary to reveal 
+a witness v1 (or greater) script-path spend, per [https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#user-content-Constructing_and_spending_Taproot_outputs BIP-0341]. This is demonstrated in detail in [[#Transaction examples|the transaction examples section]].)
+
+The items on the stack must be validated as follows:
+
+* <code><trigger-vout-idx></code> is a <code>CScriptNum</code> of up to 4 bytes.
+** It indicates the vout index of this input's corresponding <code>OP_UNVAULT</code> output.
+*** Validation rules for this output are described below.
+** If this value does not decode as a valid CScriptNum value, the script MUST fail and terminate immediately.
+** If this value does not correspond to a valid output in the spending transaction, the script MUST fail and terminate immediately.
+
+* <code><target-outputs-hash></code> is a 32 byte data push.
+** It is the hash of the proposed withdrawal target output set, defined by <code>target_outputs_hash(outputs)</code> below. Note that this value is duplicated here.<ref>'''Why, when spending an OP_VAULT output into a trigger, does the target hash need to be duplicated on the witness stack?''' The target hash exists in the <code>OP_UNVAULT</code> output script, likely behind a taproot pubkey. Its presence is required on the witness stack also so that the expected taproot pubkey for the <code>OP_UNVAULT</code> output can be constructed for comparison.</ref>
+** If this value is not 32 bytes, the script MUST fail and terminate immediately.
+
+* <code><trigger-sPK></code> is a variable length data push.
+** It must be the scriptPubKey that is the preimage to the <code><trigger-sPK-hash></code> specified in the spent <code>OP_VAULT</code> input.
+** If this value does not tagged-hash to <code><trigger-sPK-hash></code> supplied by the <code>OP_VAULT</code> parameter, the script MUST fail and terminate immediately.
+*** Verify <code>tagged_hash("VaultTriggerSPK", <trigger-sPK>) == <trigger-sPK-hash></code>
+** If this value does not correspond to a valid witness program, the script MUST fail and terminate immediately.
+
+* the remaining elements serve as the witness stack to satisfy the <code><trigger-sPK></code> witness program.
+** If <code>VerifyWitnessProgram(<remaining stack elements>, <trigger-sPK>, ...)</code> fails, the script MUST fail and terminate immediately.
+** (This validates that the withdrawal trigger has been authorized.)
+
+===== Transaction outputs validation =====
+
+Once the contents of the witness stack have been parsed and validated, the transaction outputs must be checked.
+
+First, we must check for a ''revault output'': an output in the trigger transaction whose
```

Comment:
Even this kind of address reuse has serious privacy implications. There should be a way to make it unknowable which output is change.

Reply:
You can derive a new address via key derivation and send change there. This means the user has to wait for the timelock to run out before re-spending, and at spend time it becomes somewhat obvious which is change due to timelock configuration, at a minimum.